pkg/adapters/http.go

```diff
@@ -0,0 +1,81 @@
+package adapters
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/weaveworks/weave-gitops/pkg/templates"
+)
+
+const (
+	expiredHeaderName = "Entitlement-Expired-Message"
+)
+
+// An HTTP client of the cluster service.
+type HttpClient struct {
+	baseURI *url.URL
+	client  *resty.Client
+}
+
+// NewHttpClient creates a new HTTP client of the cluster service.
+// The endpoint is expected to be an absolute HTTP URI.
+func NewHttpClient(endpoint string, client *resty.Client, out io.Writer) (*HttpClient, error) {
+	u, err := url.ParseRequestURI(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	client = client.SetHostURL(u.String()).
+		OnAfterResponse(func(c *resty.Client, r *resty.Response) error {
+			if m := r.Header().Get(expiredHeaderName); m != "" {
+				fmt.Fprintln(out, m)
+			}
+			return nil
+		})
+
+	return &HttpClient{
+		baseURI: u,
+		client:  client,
+	}, nil
+}
+
+// Source returns the endpoint of the cluster service.
+func (c *HttpClient) Source() string {
+	return c.baseURI.String()
+}
+
+// RetrieveTemplates returns the list of all templates from the cluster service.
+func (c *HttpClient) RetrieveTemplates() ([]templates.Template, error) {
+	endpoint := "v1/templates"
+
+	type ListTemplatesResponse struct {
+		Templates []*templates.Template
+	}
+
+	var templateList ListTemplatesResponse
+	res, err := c.client.R().
+		SetHeader("Accept", "application/json").
+		SetResult(&templateList).
+		Get(endpoint)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to GET templates from %q: %w", res.Request.URL, err)
+	}
+
+	if res.StatusCode() != http.StatusOK {
```

Comment:
Not really an issue with the code... just for my education: When would we git a non-ok status but no error in this context?

Reply:
This check is here more for completeness really. It's just a way to ensure that the response we got was a 200 as per the API contract. I guess a different status code could be returned without an error but that would likely be a bug on the server side. The previous error check would account for transport errors I guess.